build.sc

```diff
@@ -182,80 +126,131 @@ trait ChiselPublishModule extends CiReleaseModule {
-  def millSourcePath = super.millSourcePath / os.up / "firrtl"
+trait HasScala2MacroAnno extends CrossSbtModule {
+  override def scalacOptions = T {
+    if (!utils.isScala3(crossValue)) {
+      super.scalacOptions() ++ Agg("-Ymacro-annotations")
+    } else super.scalacOptions()
+  }
+}
 
-  override def scalacOptions = v.commonOptions ++ Seq(
-    "-language:reflectiveCalls",
-    "-language:existentials",
-    "-language:implicitConversions",
-    "-Yrangepos", // required by SemanticDB compiler plugin
-    "-Xsource:3",
-    "-Xsource-features:infer-override"
-  )
+trait HasScala2Plugin extends CrossSbtModule {
+  def pluginModule: Plugin
 
-  def osLibModuleIvy = v.osLib
+  override def scalacOptions = T {
+    if (!utils.isScala3(crossValue)) {
+      super.scalacOptions() ++ Agg(s"-Xplugin:${pluginModule.jar().path}")
+    } else super.scalacOptions()
+  }
 
-  def json4sIvy = v.json4s
+  override def scalacPluginClasspath = T {
+    if (!utils.isScala3(crossValue)) {
+      super.scalacPluginClasspath() ++ Agg(pluginModule.jar())
+    } else super.scalacPluginClasspath()
+  }
+}
 
-  def dataclassIvy = v.dataclass
+object firrtl extends Cross[Firrtl](v.scalaCrossVersions)
 
-  def commonTextIvy = v.commonText
+trait Firrtl extends CrossSbtModule with Cross.Module[String] with HasScala2MacroAnno with ScalafmtModule {
+  def millSourcePath = super.millSourcePath / os.up / "firrtl"
+  def scalaVersion = crossValue
+
+  override def scalacOptions = T {
+    utils.isScala3(crossValue) match {
+      case false => v.scala2CommonOptions ++ Seq(
+        "-language:reflectiveCalls",
+        "-language:existentials",
+        "-language:implicitConversions",
+        "-Yrangepos", // required by SemanticDB compiler plugin
+        "-Xsource:3",
+        "-Xsource-features:infer-override"
+      )
+      case true => Seq.empty[String]
+    }
+  }
+  val commonDeps = Agg(
+    v.scopt,
+    v.commonText,
+    v.osLib,
+    v.json4s
+  )
 
-  def scoptIvy = v.scopt
+  def ivyDeps = utils.isScala3(crossValue) match {
+    case true => commonDeps
+    case false => commonDeps ++ Agg(v.dataclass)
+  }
 
   object test extends SbtModuleTests with TestModule.ScalaTest with ScalafmtModule {
     def ivyDeps = Agg(v.scalatest, v.scalacheck)
   }
 }
 
 object svsim extends Cross[Svsim](v.scalaCrossVersions)
-
-trait Svsim extends common.SvsimModule with CrossSbtModule with ScalafmtModule {
+trait Svsim extends CrossSbtModule with ScalafmtModule {
   def millSourcePath = super.millSourcePath / os.up / "svsim"
 
-  override def scalacOptions = v.commonOptions ++ Seq(
-    "-Xsource:3",
-    "-Xsource-features:case-apply-copy-access"
-  )
+  override def scalacOptions = T {
+    utils.isScala3(crossValue) match {
+      case false => v.scala2CommonOptions ++ Seq(
+        "-Xsource:3",
+        "-Xsource-features:case-apply-copy-access"
+      )
+      case true => Seq.empty[String]
+    }
+  }
```

Comment:
Considering we're writing some variant of this for every single Module, I think we could simplify the typical use a bit with a common trait:
```suggestion
// Up above somewhere
trait ChiselCrossSbtModule extends CrossSbtModule {
  def isScala3: T[Boolean] = T { utils.isScala3(crossScalaValue) }
  def scala2Options: T[Seq[String]] = T { Seq.empty }
  def scala3Options: T[Seq[String]] = T { Seq.empty }
  
  override def scalacOptions = T {
    super.scalacOptions ++ (if (isScala3()) scala3Options() else scala2Options())
  }
}

  // Now using this in something like Svsim
  override def scala2Options = T {
    v.scala2CommonOptions ++ Seq(
      "-Xsource:3",
      "-Xsource-features:case-apply-copy-access"
    )
  }
  override def scala3Options = T { Seq.empty[String] }
```
One subtle useful thing here is it makes sure we include `super.scalacOptions` and makes it less likely that we accidentally forget to do that. Most of the time it doesn't matter, but it does for things using Macro2Annotations and the compiler plugin.

What do you think?

Reply:
i agree from a software perspective but i do think our future selves (and other contributors) will thank us if we optimized these build files for simple+ugly over efficient+less readable..